Multi-ControlNet Batch

ljleb opened the pull request:
Closes #243
Closes #286
Closes #654
Related #650 (request related to m2m not implemented although this is an alternative way to do it)
Related #249 (I think point 3 is not implemented here)

It is here! multi-controlnet batch, in txt2img and img2img. Batch img2img gets its own textbox:

![image](https://user-images.githubusercontent.com/32277961/228971935-acd8e595-5b2d-403c-b3e3-6407d0f271c8.png)

Each control unit can now individually specify a directory:

![image](https://user-images.githubusercontent.com/32277961/228946677-6f818423-efc6-4ebf-bf2e-bfbf9d6dfa75.png)

Still a work in progress, I need to refactor a bit.

ljleb commented:
Preview in txt2img using canny and depth on an input directory:

![image](https://user-images.githubusercontent.com/32277961/228970458-1114a8bd-23ff-4649-b22f-6d0c0c3d082d.png)

enn-nafnlaus commented:
Awesome!

Too bad I won't be able to upgrade to use it due to bug #417 .  :(

AbyszOne commented:
I hope this is successful. I still don't get why a function structurally central to the full use of the tool has been actively blocked or ignored for two months. I've been using batches with Comfy for a while now because I've given up hope here, when even amateur scripts have added this feature long ago.
About point 3, it requires either an adaptation of alternative test, or training a controlnet for that function, so we could ignore it for now.

ljleb commented:
> I still don't get why a function structurally central to the full use of the tool has been actively blocked or ignored for two months.

It doesn't seem like anybody wanted to write the code for batch processing inside the extension for a while, so I just decided to go for it on a whim this morning.

I am not sure what you are referring to by blocked? Although I understand your frustration with this. No one is forced to contribute, no one is paid to work on this project after all. This is open-source.

ljleb commented:
BTW I was thinking of adding support for TemporalNet after this PR if no one wants to take it.

AbyszOne commented:
The original PR to include batch in the GUI was not accepted by Automatic. And after being requested countless times, this feature hasn't been added to the interface for months now. That, in my dictionary is to ignore. I'm not saying it's mandatory, I'm saying it's been actively ignored. This is so.
TemporalnNet is new to me. Unfortunately, I don't see how that improves the traditional methods. My script reduces flickering much more efficiently, and I still don't consider it optimal. Perhaps the examples given do not do justice to the method.

AbyszOne commented:
> After a bit more testing, the code seems to break multidiffusion at the moment. I don't know how it interacts with controlnet so this may take a while to fix.

Perhaps you may check the controlnet toyxyz extension for multi batch script. https://toyxyz.gumroad.com/l/jydvk

ljleb commented:
> @AbyszOne Perhaps you may check the controlnet toyxyz extension for multi batch script. https://toyxyz.gumroad.com/l/jydvk

Implementing this as a script is more trivial than what I am trying to do. This PR is not a script, which means we cannot call sample multiple times for example like the XYZ grid script does.

ljleb commented:
It was relatively trivial in the end. It seems I could simply rely on `script.process_batch`. Will refactor a bit, add a couple tests and then mark as ready.

ljleb commented:
I am not sure what the best place for the batch fallback textbox is. Right now the one in img2img batch tab is always used, even in txt2img, when leaving the individual batch textboxes empty.

AbyszOne commented:
Good news. If this progresses, it would be interesting to raise some of these multi layer/folder optimizations discussed in previous threads: https://github.com/Mikubill/sd-webui-controlnet/issues/243#issuecomment-1447603112_

However, while that's important at scale, I think a relatively simple functionality to add would be a loopback. That is, that the generated image is used as a base in Img2Img for the next generation. This approach is used in several custom scripts, as it allows to considerably increase time consistency in combination with color correction. A native implementation would be highly beneficial.

ljleb commented:
> @AbyszOne However, while that's important at scale, I think a relatively simple functionality to add would be a loopback.

That's a good idea. TemporalNet requires a similar approach so I will try to fit the last-frame loopback with the next PR.

ljleb commented:
I'm adding 1 or 2 more tests and I think this would be ready to merge. It's hard to automatically test many important part of the code that also loads models, AFAICT we would need to hijack a lot (or heavily refactor) `process` to get it to work.

ljleb commented:
I decided to put the loopback feature into this PR as well because I ended up refactoring the img2img batch detection code into something more reliable. If this makes the PR too big, I'll redo the commits so that it's easier to review.

Changelogs:
- added multi-controlnet batch, see screenshots above for how it was designed
- added checkbox to feed last generation back into the preprocessor when in batch-controlnet mode
- replaced the img2img tab tracker with a hijack on the `img2img.process_batch` function
- to implement the batch code, I had to relocate the detectmap conversion code into `compute_control_map`. This function does not do anything new
- created a new class `UiControlNetUnit` to distinguish between ui units and external code units. The batch code in this PR only works with the ui. If you want to batch process from the API or external code please call the script as many time as you need while orchestrating your requests.
- now all external code units (both web api and script api) are normalized into batch sizes of 1, to simplify further processing. See `normalize_to_batch_mode` for the implementation. This function is also unit tested
- move preprocessor application to `before_process_batch` and `postprocess_batch`. This means each image will be processed one by one instead of all at once in `process`

Let me know if anything needs to be changed or if anything is unclear, I'll try to do my best to fix and clarify these.

ljleb commented:
The new checkbox was put here:

![image](https://user-images.githubusercontent.com/32277961/229273298-a2a9f955-1cea-4387-b85d-f860e3882255.png)

I am not sure if it has a place in batch mode as it will only consider the first image of the batch. I thought it would be useful to use the first image when using a shared batch folder. I think TemporalNet would be slightly easier to use like this.

AbyszOne commented:
Cool work, but the loopback is the other way around. You need to send the loopback to Img2Img, no to CN. I mean, that can be somewhat useful, but not the method I was talking about.

ljleb commented:
Ah indeed I misinterpreted your comment. So IIUC, instead of sending the last generated image to a cn unit, you want to send it to the img2img init image.

TemporalNet needs the chackbox I added to work with other types of control models, so the checkbox does have its purpose.

In fact, this would be useful with or without controlnet I think. We may want to look into making a PR in the webui repo instead for the img2img init image loopback.

It could also be implemented as an alwayson script that adds a checkbox in the img2img tab. If we implement it as a separate script, we need to be able to install it without installing cn, but also install it automatically when installing controlnet because we dont want to have to install multiple extensions for the full cn experience.

AbyszOne commented:
Yes, I think both options are useful. I just hope they could be integranted as batch options in the CN extension without rest on main UI PRs.

ljleb commented:
After trying to implement the img2img init image loopback as a script, it seems it won't be possible to do that, as `processing.StableDiffusionProcessingImg2Img.init` processes them in bulk at the beginning of the sampling process. Will need a PR in the webui repo.

The reason it is possible to implement this as a selectable script is that you have control over how many times you invoke the sampling process.

AbyszOne commented:
It escapes my comprehension. I can only hope that it can be implemented, somehow.👍

ljleb commented:
In the end I found a way to make img2img loopback work. My mistake was mixing up batch size with batch count.

However, I encountered a problem while implementing it. There seems to be 2 different kinds of batch modes in the webui: img2img batch and simple batch processing.

For example, when processing a directory of 5 images in img2img batch + using the XYZ script to generate a grid of 1x3 + setting batch count = 2, it will:
- enter `modules.img2img.batch_process`
- call the XYZ script 5 times
- each time, XYZ script generates 1 images grid of batch count * batch size * grid size

The implementation I opted for increases the batch count, but it should instead try to use the `img2img.batch_process` function in some way. It is otherwise impossible to run the img2img init image loopback inside the img2img batch tab in a coherent way.

AbyszOne commented:
Any news?

ljleb commented:
I took a deep dive these last 2 days into making img2img loopback work as a standalone script, you can find my efforts here: https://github.com/ljleb/sd-webui-controlnet/tree/img2img-loopback

I was looking into how to make it general. I found a pretty neat loopback script that is not focused on movie2movie but still has some importance I think: https://github.com/Extraltodeus/advanced-loopback-for-sd-webui

I am not sure how we should implement the img2img loopback. As a selectable script, you cannot use the alternative img2img test script (or any other script). As an alwayson scipt, it could make sense to put it into its own extension; however, doing this means you need multiple extensions to get the full controlnet m2m features.

So I was thinking of making a PR with this new script added to this repo, even though in theory you could make it into its own extension.

What do you think, is this a bad idea? Make loopback its own extension or put it here?

sssttttas commented:
I'm trying to reuse an image sequence that has been preprocessed with controlnet beforehand. I'm feeding the path to the Controlnet input directory on the batch tab. Then select "none" (because it's already preprocessed) in controlnet's preprocessor dropdown menu.

As the result - I'm not sure what it does exactly - but it seems that CN ignores the preprocessed sequence and uses original image sequence as detectmap. 

I'm I misundestanding something? I thought using preproccessed sequences in batch processing is the why we need batch input for control net in the first place.

ljleb commented:
> As the result - I'm not sure what it does exactly - but it seems that CN ignores the preprocessed sequence and uses original image sequence as detectmap.

Thanks for testing this. At the moment, a number of aspects need to be worked on and I am looking into making the code work as it should. This is still a work in progress, not at all ready for review. I do not want to waste your time, I'll mark as ready for review when I think it needs to be tested.

sssttttas commented:
I'm just can't wait for it, I'm having  hanging project heavely depending in CN batch processing. Thank you for your work!

ljleb commented:
I know this is an important feature... To be honest I should be working on my personal duties instead of this PR 😅 I'm trying to go as fast as I can without rushing things.

If you need a quick working solution, You may want to look into @AbyszOne project or check out some scripts made by the community i.e. @toyxyz batch script.

ljleb commented:
Since the commits were all over the place, I reordered the changes into something more cohesive. Hope this helps.

If I did not forget anything, this should be relatively complete now. I may have missed one detail or two, IMO this should be tested by 1 or 2 other people before merging to make sure everything works as intended. Any feedback on the code is appreciated.

✅ Implemented in this PR:

- specify a batch directory for each active controlnet model
- you can mix units in single image mode with other batch units. Single images in "upload" mode are reused as many times as needed for all batch iterations. The length of the smallest batch is the length of the entire batch (additional images in other batches are skipped)
- "img2img batch processing" and "batch count/batch size" are distinct concepts in the webui, controlnet batch hijacks the webui to fit itself in the former category (supports both txt2img and img2img)
- save processed images into specified batch directory if desired (specified in the img2img batch tab) and otherwise save in your webui images directory
- individual cn model inputs all fall back on the master controlnet input directory (you can find this new textbox in the img2img batch tab), which itself falls back on the existing img2img batch input directory
- multi-controlnet loopback: put last generated image into this unit. This is only active during batch processing and if selected. It will not read the last generated image from a previous batch or terminated generation. This allows us to support TemporalNet
- 2 new options:
   - "Show batch images in gradio gallerie output": whether to show images in the gradio interface after generation. Do not enable this if you are processing a lot of images
   - "Increment seed after each controlnet batch iteration": while implementing img2img loopback, I found that keeping the same seed makes the output increasingly burned and high-contrast. This will come in handy when you need a different seed for each batch iteration
- a number of bug fixes: namely the img2img batch tab tracker has been replaced with the monkey patch object that enables controlnet batch processing, and IIUC preprocessors unloadable callbacks were never really called.

⛔ Not implemented in this PR:

- img2img loopback (loopback into the img2img init image)
- different merging strategies and quantity of merge for loopback. (whether for img2img init image or multi-controlnet) We may want to take inspiration from toyxyz's script for the different merge strategies. I have been exploring alternative ideas for loopback merge ratios, however there isn't enough space to cover everything here. In any case, after fiddling a bit with it, it seems to me that img2img loopback has a lot of potential
- txt2movie: generate n consecutive images in txt2img without an existing directory of images to process. We may want to create a selectable script for this. To support other selectable scripts simultaneously, we could add a dropdown in this script to use a nested script, so that for example we can use img2img mk2 with it.

sssttttas commented:
I tested the workflow I was taking about before. Unfortunately it's still not working. 
Just in case I didn't express myself clearly. I'm trying to feed into cn a sequence of normal maps I got from a 3d software.  To do this I'm selecting "preprocessor - none" and "model - control_normal-fp16 [63f96f7c]". Leaving the upload field emply, putting path to the sequence in ControlNet input field.

When I run it on a sigle image it works fine (normal map uploaded manually). 
When runned in batch it makes something completely different (a total mess).

ljleb commented:
Interesting, on my side it seems to work with open pose for example. I am using diff models, maybe it has to do with complete models being reapplied every time without being reset, or something like that?

ljleb commented:
So I tried reproducing your setup @sssttttas but unfortunately I could not reproduce. Here's what I did:

1. Go to img2img batch tab
2. enter both output and controlnet-input dirs
3. go to txt2img tab
4. pick settings
5. open controlnet txt2img accordion
6. enable model 1
7. click on batch tab button for model 1 to enter batch mode
8. pick model "control_sd15_normal [e3b0c442]" (not diff nor fp16)
9. click on generate. I see "controlnet batch mode" in console, which tells me the batch hijack code was successfully patched. Also n images are generated in the output dir picked above

I also tested this with model "diff_control_sd15_normal_fp16 [00173cc1]", which also worked. Let me know if I did not understand your setup properly, or if you did something slightly differently. If it is different, please share a screenshot of your settings so I can reproduce in the same environment.

Order of input images is order of output images:

<details>

<summary>Input Images</summary>

![NormalMap1](https://user-images.githubusercontent.com/32277961/231026541-730fcfaa-e8eb-4a58-90f9-c42388e766bf.png)
![NormalMap2](https://user-images.githubusercontent.com/32277961/231026543-193175aa-caef-478d-8241-a908495d8f5a.png)
![NormalMap3](https://user-images.githubusercontent.com/32277961/231026545-dc5e33cf-6cdd-46c8-ad66-774433d231ad.png)
![NormalMap4](https://user-images.githubusercontent.com/32277961/231026547-cd7d9400-004e-4281-a94e-5c8073a8703d.png)
![NormalMap5](https://user-images.githubusercontent.com/32277961/231026550-04e7c25c-3216-4d21-9624-73cbe54145bd.png)
![NormalMap6](https://user-images.githubusercontent.com/32277961/231026552-84dffb82-c738-4220-876a-250aadf3b29a.png)
![NormalMap7](https://user-images.githubusercontent.com/32277961/231026555-8ee092a2-f263-4ac8-b063-4cf665cb164a.png)
![NormalMap8](https://user-images.githubusercontent.com/32277961/231026557-78960223-fe8f-406c-b6ac-344456ca698e.png)

</details>

<details>

<summary>Output Images</summary>

![NormalMap1](https://user-images.githubusercontent.com/32277961/231026600-71f96c1a-6722-4f5a-b63c-e7d00180f2b5.png)
![NormalMap2](https://user-images.githubusercontent.com/32277961/231026601-e05ab16c-edbc-4bb2-b072-c1d88c1666af.png)
![NormalMap3](https://user-images.githubusercontent.com/32277961/231026602-9de0fd35-4c62-41a3-b83b-5a867ddf8ab8.png)
![NormalMap4](https://user-images.githubusercontent.com/32277961/231026604-1942f170-d3ab-460d-b283-74faeaea208c.png)
![NormalMap5](https://user-images.githubusercontent.com/32277961/231026605-a5d8e780-07c1-40a2-b457-6cf9a0f380c1.png)
![NormalMap6](https://user-images.githubusercontent.com/32277961/231026608-52db6cd9-4c9b-4db8-b03b-52d87532fe4e.png)
![NormalMap7](https://user-images.githubusercontent.com/32277961/231026614-31196373-bde6-4aa2-b691-9cae6869cd25.png)
![NormalMap8](https://user-images.githubusercontent.com/32277961/231026615-fa993cb3-7141-42db-a5e9-44a87e71668d.png)

</details>

sssttttas commented:
Oh, thanks for your process breakdown. I should've started with sharing mine, but it didn't cross my mind we could use it differently. I'm trying to use cn to keep objects recognizible while applying a style to an original img sequence. In my case I'm applying painterly look to a 3d scene. Here's the workflow:

Prepearing part
1. img2img tab, uploading 1 frame, entering promt, picking settings
2. Enabling cn models (usually hed, normal or depth). Tested with one or more, cn count doen't make difference.
3. Uploading accorging preprocessed frames.
4. Setting "preprocessor" to "none".
5. Test run with one frame.

Batch processing
1. Switching to img2img batch tab.
2. Input directory -- path to original img sequence
3. Output directory -- can be specified or left empty
4. Controlnet input directory -- tried leaving empty or put path to the input sequence for the 1st cn.
5. Going to each of the CNs, removing uploaded signe frames, putting paths to the corresponding preprocessed sequences  in  input directory fields.
6. Leaving preprocessor and model settings as in Prepearing part. 
7. Running batch.
8. Comparing results from the test signle frame run and 1st framу of the batch processing.
9. Crying :(

My test sequences: https://disk.yandex.ru/d/aM4XwLhKLNUIRQ
Video of the process: https://disk.yandex.ru/i/LiWu-yHaU_K7DA

Single run result (desired outcome):  https://freeimage.host/i/HkDT1rG
First frame from batch processing: https://freeimage.host/i/HkDIgWX

ljleb commented:
@sssttttas Thanks for your detailed step-by-step reproduction guide. I could reproduce, and hopefully fix the issue as well. Let  me know if it works now on your side.

Enferlain commented:
I read through the thread but I don't get it. If I have 200 frames in a folder, and I want to batch process them to get the detectmap of each frame to reuse for a video, does this help? I tried just normal i2i batch with controlnet on and auto save detectmap but for openpose all the saved detectmaps are just black. With depth it worked though, not sure what makes them different.

ljleb commented:
> for openpose all the saved detectmaps are just black

And they are not black if the preprocessor is ran on its own manually, correct? I personally get black images with openpose sometimes, because the preprocessor does not know how to deal with certain types of situations.

To run preprocessor only, you could go in img2img batch mode and set denoising strength to 0, it should basically only run the preprocessor. Not ideal but it should do the job.

Enferlain commented:
Well yeah, if I don't have a controlnet model selected it doesn't do anything

```
Traceback (most recent call last):
  File "/content/drive/.shortcut-targets-by-id/1L0fJf5wzsbvNAkNMlHU_amjVtMuFfIUR/stable-diffusion-webui-colab/stable-diffusion-webui/modules/scripts.py", line 409, in process
    script.process(p, *script_args)
  File "/content/drive/.shortcut-targets-by-id/1L0fJf5wzsbvNAkNMlHU_amjVtMuFfIUR/stable-diffusion-webui-colab/stable-diffusion-webui/extensions/sd-webui-controlnet/scripts/controlnet.py", line 679, in process
    model_net = self.load_control_model(p, unet, unit.model, unit.low_vram)
  File "/content/drive/.shortcut-targets-by-id/1L0fJf5wzsbvNAkNMlHU_amjVtMuFfIUR/stable-diffusion-webui-colab/stable-diffusion-webui/extensions/sd-webui-controlnet/scripts/controlnet.py", line 468, in load_control_model
    model_net = self.build_control_model(p, unet, model, lowvram)
  File "/content/drive/.shortcut-targets-by-id/1L0fJf5wzsbvNAkNMlHU_amjVtMuFfIUR/stable-diffusion-webui-colab/stable-diffusion-webui/extensions/sd-webui-controlnet/scripts/controlnet.py", line 483, in build_control_model
    raise RuntimeError(f"model not found: {model}")
RuntimeError: model not found: None
```

The 0 denoise idea really helps though, makes things a lot faster, so thanks. I just wish I could get openpose detectmaps as well

Depth seems alright

![image](https://user-images.githubusercontent.com/15861396/231553985-9e2178d7-9b80-4f37-9a77-f4d6112070b8.png)


Also canny does this nightmare thing

![image](https://user-images.githubusercontent.com/15861396/231553894-8a94f396-1c2f-4b4b-990d-1eb4559bf98d.png)

FurkanGozukara commented:
lol I was looking for this gonna test now

FurkanGozukara commented:
@ljleb 

it is generating first image then stops any ideas? it just waits forever no messages

using latest automatic1111 web ui and controlnet. merged your pull request

here all the settings

![image](https://user-images.githubusercontent.com/19240467/231756511-a8c4e797-4222-42b9-827d-1003265cef34.png)

![image](https://user-images.githubusercontent.com/19240467/231756666-0fe4993d-069b-4825-adc3-efd465f7c5d5.png)

![image](https://user-images.githubusercontent.com/19240467/231756713-fe1aed8f-7d10-4581-bc63-8fbfa7dfedb5.png)

sssttttas commented:
@FurkanGozukara  Try putting path in batch tab in the controlnet as well.

ljleb commented:
Maybe the ui needs a bit of work? When you are in the batch tab, you put the unit in batch mode, and when you are in upload tab, you put the unit in single image mode, which is useful if you want to use the same control image for the whole batch.

ljleb commented:
> @Enferlain Also canny does this nightmare thing

Is this only happening using the code in the PR or is it related to code already present in main? Not sure which part of the code I need to look at to reproduce.

sssttttas commented:
> Maybe the ui needs a bit of work? When you are in the batch tab, you put the unit in batch mode, and when you are in upload tab, you put the unit in single image mode, which is useful if you want to use the same control image for the whole batch.

I would be more obvious if you replace tabs with a radio button. A tab feels like it should work even if it's closed.

ljleb commented:
> I would be more obvious if you replace tabs with a radio button. A tab feels like it should work even if it's closed.

I understand. I know how it's implemented in the background, so it's harder for me to say. I'll see what I can do.

The idea was to use the same design as the img2img subtabs, i.e. the selected tab is the active type of processing. For example, when you go to img2img sub tab, you can only inpaint the whole image; when you go to the inpaint subtab, you can only inpaint using a mask; when you go to the batch subtab, you can only batch process multiple images from an input directory.

I was thinking we may want to split uploading images from sketches, maybe with more than just 2 tabs it would be more obvious. Would you say we're in a different situation here or it's just the small number of tabs that makes it less obvious?

Enferlain commented:
@ljleb This was using controlnet on batch. Select preprocessor and model to canny and then watch the depth maps turn into cthulhu in real time. Depth also seems much heavier and has rougher outlines when result of batch processing. And openpose was black image as mentioned previously.

Although today I updated to this -- e5b565e2 (Thu Apr 13 16:38:14 2023) and I haven't checked yet. I'll try it a bit later.

ljleb commented:
> @Enferlain This was using controlnet on image2image batchThis was using controlnet on image2image batch

IIUC this is unrelated to this PR, correct? If it is unrelated, it means we need to fix in another PR. We do not want to block the merge of this feature for an orthogonal concern.

Enferlain commented:
@ljleb Yeah sorry, it's probably not related to this. I'll look into it a bit later and make a thread about it to try and figure out what's happening.

ljleb commented:
Ah no problem. I hope I didn't come out as a little bit rude above 😅 In any case, thanks for your time! I'll lend a hand when I find more time.

milosleeps commented:
I really hate to ask this. Could someone point me in the right direction as to how to install this?

ljleb commented:
> I really hate to ask this. Could someone point me in the right direction as to how to install this?

Before anything, you may want to know that only a couple people tested this. It should work, but please let us know here if you find any problem or have feedback.

First, you need to open a terminal in `<your webui install>\extensions\sd-webui-controlnet`. To do that:

1. navigate to your sd-webui-controlnet install location in a file explorer
2. while holding shift, right click and select "open in powershell"

A terminal window should open at the right location. Once this is done, you can merge using the method github recommends:

```
git fetch origin pull/683/head:batch
git checkout batch
```

Now, you can just reload the gradio interface, or start your webui if it was closed.

You can revert this change by typing in this terminal or a new one (opened in your sd-webui-controlnet install) at any time:

```
git checkout main
```

sssttttas commented:
@AladdinFPV 
It increments the seed with each frame by default. Too keep the seed uncheck "Increment seed after each controlnet batch iteration" in settings>controlnet. 

Having an incremented seed is useful in cases when image is changing fast (for example, a video with a fast moving camera).

ljleb commented:
> the image is really different in each generation, as if it's ignoring the seed?

It's as @sssttttas, although I think it should be turned off by default actually. It's useful for loopback stuff, as img2img init image loopback with the same seed generates some pretty heavy artifacts.

AladdinFPV commented:
> > the image is really different in each generation, as if it's ignoring the seed?
> 
> It's as @sssttttas said, although I think it should be turned off by default actually. It's useful for loopback stuff, as img2img init image loopback with the same seed generates some pretty heavy artifacts.

Hi thanks a lot, It makes sense but doesn't seem to be working, do I need to keep "Do not append detectmap to output" and "Passing ControlNet parameters with "Send to img2img"" both checked?

I also tried removing the init photo and CN photo in img2img before switching to batch and same results.

these are my current settings

![image](https://user-images.githubusercontent.com/130452224/232458519-bde51632-8a58-4d4d-acf5-2b91784da4fe.png)

![image](https://user-images.githubusercontent.com/130452224/232458586-81542787-fb4b-4ed8-9362-e19b0553fb55.png)

![image](https://user-images.githubusercontent.com/130452224/232458704-8c4a241f-bc03-40c2-8800-34a9c5cd0512.png)

ljleb commented:
Can you clarify what doesn't seem to be working? I am not sure I understand what you expect to see vs what you get. If you could share an expected result vs what it actually does, it would help with resolving the issue.

AladdinFPV commented:
> Can you clarify what doesn't seem to be working? I am not sure I understand what you expect to see vs what you get. If you could share an expected result vs what it actually does, it would help with resolving the issue.

Sorry I didn't explain enough, I was following the thread from what @sssttttas was facing here https://github.com/Mikubill/sd-webui-controlnet/pull/683#issuecomment-1503349072 

Let me explain my situation in detail:
input (1 frame out of 400)
![BK0000](https://user-images.githubusercontent.com/130452224/232463601-6c83a9db-aa28-4cdb-a07d-3dcb15e3de2b.png)

output
![00161-3610865726](https://user-images.githubusercontent.com/130452224/232463641-7b2e3425-181a-4b70-b514-e658d075fe68.png)

first issue faced that I was unable to reproduce the same exact output image even thought I used the same exact details and seed and AFAIK it should produce the same image.
I used PNG info tab to double check but didn't get the same results still
![image](https://user-images.githubusercontent.com/130452224/232464288-e8a91f9a-27f1-41a5-a681-75b3cf3f3ead.png)

parameters

"(cyberpunk city), ((Hyperdetailed)), (masterpiece), 8k, (((ultra realistic))), futuristic, ((colorful image)), ((bright image)), ((LED displays covering all buildings)), (((bright sky))), ((Orange/golden moon with anamorphic flare))
Negative prompt: wires, cables, power lines, text, smooth, reflections, fireworks, duplicate buildings
Steps: 45, Sampler: Euler a, CFG scale: 11, Seed: 3610865726, Size: 576x1024, Model hash: ffad8f3da7, Model: DreamShaper_5_beta2_BakedVae_fp16, Denoising strength: 0.76, ControlNet Enabled: True, ControlNet Module: depth, ControlNet Model: control_sd15_depth [fef5e48e], ControlNet Weight: 0.9, ControlNet Guidance Start: 0, ControlNet Guidance End: 0.85, Noise multiplier: 1.05"

Image was produced by having this photo in the init Img2img and original input photo in controlnet photo.
![f01a4cd9-72c4-46fb-811f-a66898cf8cf4233](https://user-images.githubusercontent.com/130452224/232464724-27485fab-79ac-4644-b5eb-61e307b1522e.png)

This is still okay since I can just generate another style or something similar then use fixed seed.

Moving on to batch from this point, I want to generate the same output for all 400 images, since the movement is very slow the 400 images are almost identical.

My process from here was go to batch in img2img tab > select input and out put directory and left controlnet directory empty to take from input directory, and make sure same seed is selected as per the last image produced.
In "controlnet 0" I selected batch and left the directory empty to take input as per img2img controlnet directory

selected model control_sd15_depth.pth and didn't select Preprocessor. (tried to select Preprocessor as depth_midas as well)

my expected outcome is to have the 400 images with the same output image produced in img2img tab while reading the depth controlnet from every image in the input folder to apply the slight movement for each frame, in order to produce a moving video.

Thanks a lot I've been trying to do this for like 4 days straight and can't seem to make it work.

ljleb commented:
Thanks for the detailed explanation and images, it helps a lot. If I had to make a quick guess, the moon in the second example clearly is stretched. Maybe it is related to selected stretch mode? What happens if you pick the same stretch mode in both controlnet and img2img settings? i.e. both "Just resize", or both the middle option.

ljleb commented:
Thanks, should be fixed now.

AladdinFPV commented:
> 

I usually match both resize, I use resize only.

The moon in the 2nd image is just a reference image for img2img to apply style for the input image which I put in controlnet.

In batch mode, every single output is coming different even with the same seed and settings.

If I'm understanding right, the init image in img2img is to apply style?
and controlnet image is the input, the structure and depth of the image that I want to maintain

Basically what I'm trying to do is very similar to what's in this video but to do it in batch to apply it to a video.
https://www.youtube.com/watch?v=wK2vMlHREqs&ab_channel=CHILDISHYT

ljleb commented:
Ah I thought you had shared both the previous output you had and the new one using this branch, my bad. What does the new image look like?

AladdinFPV commented:
> Ah I thought you had shared both the previous output you had and the new one using this branch, my bad. What does the output look like using the code in this branch?

![00150-3610865726](https://user-images.githubusercontent.com/130452224/232478568-55a8fd6e-83ea-460d-92df-9393109ccccd.png)
this was generated with the same settings and seed and inputs.

ljleb commented:
@AladdinFPV I found your model and tested with the available information. I could not replicate exactly your images, however I could reproduce your issue:

<details>
<summary>Output on main</summary>

![00001-3610865726](https://user-images.githubusercontent.com/32277961/232545370-fde31160-f2d6-46c1-b706-312934a90333.png)

</details>

<details>
<summary>Output on 0f549888fd49aea48a4a5049f75c2e87ad3affad</summary>

![00002-3610865726](https://user-images.githubusercontent.com/32277961/232545466-aae09a66-5f15-4e76-b278-874deabc0473.png)

</details>

So the issue appears to be unrelated to this PR. We may need to git bisect the 1.1 update to see what causes this.

ljleb commented:
I think we should meege this when 1.1 is more stable. Otherwise we will be adding complexity to the code and I am pretty sure I missed 1 or 2 details that we will have to fix after the merge.

marcsyp commented:
Is there a way I can test this before it is merged?  I have not tried to check out a PR yet, how is it done properly?  Or should I just wait?  Thanks!

jjhaggar commented:
> Is there a way I can test this before it is merged?

Yeah, you can follow these steps to try it before it's merged: https://github.com/Mikubill/sd-webui-controlnet/pull/683#issuecomment-1510475436

FurkanGozukara commented:
> I think we should merge this when 1.1 is more stable. Otherwise we will be adding complexity to the code and I am pretty sure I missed 1 or 2 details that we will have to fix after the merge.

when will be 1.1 become stable? i am planning hopefully a new tutorial for 1.1

ZZZephyriusss commented:
> > I really hate to ask this. Could someone point me in the right direction as to how to install this?
>
> First, you need to open a terminal in `<your webui install>\extensions\sd-webui-controlnet`. To do that:
> 
>     1. navigate to your sd-webui-controlnet install location in a file explorer
> 
>     2. while holding shift, right click anywhere in the file and folders area and select "open in powershell"
> 
> A terminal window should open at the right location. Once this is done, you can checkout the code using the method github recommends:
> 
> ```
> git fetch origin pull/683/head:batch
> git checkout batch
> ```
>

For the newbies like me, if shift + right click doesn't work, just tape inthe explorer in the right folder : CMD and it will open the console the merge the PR.

I have tested it with a folder full of 100+ png for openpose models and it works so fine !!! So easy now to test a chechpoint model with multiple poses in one click !! Thank you so much !!

ZZZephyriusss commented:
Bug Report

Hello ! Thanks for your work !! I have beeing testing it and it works fine but if i have a folder inside the input folder it crashes..or it looks like it is the cause. I could go back to the main branch, Stable diffusion doens't start at all now, i have this message : 


"Commit hash: 64b7e8382377bb578d9740c061979776b214cfd9
Traceback (most recent call last):
  File "C:\AI FACTORY\NEW SD\stable-diffusion-webui\launch.py", line 380, in <module>
    prepare_environment()
  File "C:\AI FACTORY\NEW SD\stable-diffusion-webui\launch.py", line 287, in prepare_environment
    run_python("import torch; assert torch.cuda.is_available(), 'Torch is not able to use GPU; add --skip-torch-cuda-test to COMMANDLINE_ARGS variable to disable this check'")
  File "C:\AI FACTORY\NEW SD\stable-diffusion-webui\launch.py", line 137, in run_python
    return run(f'"{python}" -c "{code}"', desc, errdesc)
  File "C:\AI FACTORY\NEW SD\stable-diffusion-webui\launch.py", line 113, in run
    raise RuntimeError(message)
RuntimeError: Error running command.
Command: "C:\AI FACTORY\NEW SD\stable-diffusion-webui\venv\Scripts\python.exe" -c "import torch; assert torch.cuda.is_available(), 'Torch is not able to use GPU; add --skip-torch-cuda-test to COMMANDLINE_ARGS variable to disable this check'"
Error code: 1
stdout: <empty>
stderr: Traceback (most recent call last):
  File "<string>", line 1, in <module>
AssertionError: Torch is not able to use GPU; add --skip-torch-cuda-test to COMMANDLINE_ARGS variable to disable this check"

not sure if i can do something or if i have to reinstall Sd...!

ljleb commented:
Hmm that would be surprizing. The code that lists files filters anything that is not a file:

https://github.com/AUTOMATIC1111/stable-diffusion-webui/blob/master/modules/shared.py#L643

lllyasviel commented:
I do not know what it is and how it works. but you just merge if no bugs can be detected. if changes are super large perhaps we take a look

ZZZephyriusss commented:
> Hmm that would be surprizing. The code that lists files filters anything that is not a file:
> 
> https://github.com/AUTOMATIC1111/stable-diffusion-webui/blob/master/modules/shared.py#L643
> 
> Strange issue. Could you try to reproduce it to be sure it is related to the image loading code from this PR?

UPDATE : i tested it again with subfolders and others none related  formats inside the input folder for ControlNet : No probleme...

I wil use intensively your pull in the next week and report anything abnormal. Thanks you so much for your contribution, it help so much !!!

sssttttas commented:
@ljleb I was wondering, would it be possible to release the **incrementing batch seed as a separate extension/script?** 

May be it's not a very frequently requested feature but I was looking for it for quite a while. And since you've already implemented it your PR I hope it will not be too much hassle to make a separate extension out of it. Thank you!

ljleb commented:
> would it be possible to release the incrementing batch seed as a separate extension/script?

This would be a quite small feature to put alone in a different script. Ultimately this can be useful for img2img init image loopback IIUC, which I believe we want to support in this extension.

I will soon be done with resolving the conflicts and then testing that the changes from main were incorporated properly. Once this is done and all known issues have been resolved, I'll merge this into main.

sssttttas commented:
> This would be a quite small feature to put alone in a different script. Ultimately this can be useful for img2img init image loopback IIUC, which I believe we want to support in this extension.

Will I be able to use incrementing seed feature alone (without loopback or CN)? I need it with the regular batch img2img processing. Maybe it would even make sense to include it in the Automatic1111 webui itself?

ljleb commented:
Ah wait you're right, it only activates when controlnet is in batch mode, not when the webui is in batch mode.

Sure I can make an extension. There are multiple options for loopback at this moment, but I can try to put it into a custom loopback script. I am not sure how I feel to require people to install multiple extensions to get the full img2img loopback experience. I guess you can always drop my script if it ends up making sense to integrate it here, and otherwise it will be its own thing.

ljleb commented:
All known issues related to this PR have been fixed. A number of issues will probably still emerge from this change, I'll be on my guard for the next day so that obvious mistakes are sorted out quickly.

ljleb commented:
@sssttttas https://github.com/ljleb/sd-webui-batch-loopback

For now it only includes the seed increment feature per your request. I will add more feature in a while. In the next week I will not be able to touch a single line of code unfortunately.

I tested the code, it should work with controlnet batch, as well as with the img2img batch tab.

Feel free to open an issue or PR if you have additional requests.

sssttttas commented:
@ljleb It works great, thank you!

haydenmartinGAN commented:
I absolutely love this new feature but there’s a few issues when it comes to using scripts alongside the batch controlnet feature that I want to highlight in an effort to help progress

I’m not a coder at all, so I’ve tried to make a decent visual representation of the issue...

Man, it’s so annoying how I can’t just run multiple scripts at the same time. I’ve tried merging multiple script files into one but just don’t have any luck. I’m using batch img2img trying to run Loopback with Controlnet and Prompt from file but the interplay is fucked 

—-

ok i'll try and make this straightorward

right now i need to run the batch img2img tool (found by selecting the img2img, then selecting the Batch tab)

if you scroll down to scripts, you will see a script called "Loopback", and another called "Prompts from file or textbox".

Loopback - sends an img2img generation back into itself, essentially reapplying the prompt to itself in a feedback loop - particularly helpful when used with controlnet to really get a character into a pose without too much artefacting.

Prompts from file or textbox - provides you the ability to chronologically run through a list of prompts. when used with img2img, it take input image 1 and applies however many prompts you have listed (lets say 5), so input image 1 becomes 5 output images.

What I really want is for both of these scripts to work together ideally sequentially. Here's how i want it to work:

img2img batch mode:

INPUT IMAGE 1 > Prompts from file or textbox (5 prompts)> APPLY PROMPT 1 > LOOPBACK WITH PROMPT 1
INPUT IMAGE 2 > Prompts from file or textbox (5 prompts)> APPLY PROMPT 2 > LOOPBACK WITH PROMPT 2
INPUT IMAGE 3 > Prompts from file or textbox (5 prompts)> APPLY PROMPT 3 > LOOPBACK WITH PROMPT 3
INPUT IMAGE 4 > Prompts from file or textbox (4 prompts)> APPLY PROMPT 4 > LOOPBACK WITH PROMPT 4
INPUT IMAGE 5 > Prompts from file or textbox (5 prompts)> APPLY PROMPT 5 > LOOPBACK WITH PROMPT 5

I’ve tried to make some images that explain the problem.

Any devs out there wanna help me out here?

![22E9C723-E512-47A2-8C11-9AF393204FCD](https://user-images.githubusercontent.com/49661208/235957521-db2294f0-8a63-4d50-82ec-1782ba9e20b8.jpeg)
![730D88F3-116E-45C8-8D0A-0829729E1A2A](https://user-images.githubusercontent.com/49661208/235957535-57e4b590-e43e-45c6-826b-62005dad2b40.jpeg)

ljleb commented:
The intent behind the controlnet loopback checkbox was to make it possible to use the temporal controlnet model which you can find here: https://huggingface.co/CiaraRowles/TemporalNet

As for the incompatibility between different scripts for batch, it stems from the different batch sequence strategies provided by the ui and the limitations of selectable scripts. This will require a redesign in a1111's webui to make sane and possible the interaction between different extensions and scripts that provide batch support.

In the meantime, your best bet is probably to create your own script to schedule the generation.

zhoobin021 commented:
Thank you to the amazing souls who made this possible, you guys are awesome. 🌸💜🙏

It's also amazing that we have multi CNs, each with its own batch inside.
This way we can have multiple images influencing on top of an image, also the other way around we can use batch in img2img to give different styles to a single image, nevertheless it's also super useful for video(img sequence).

I didn't understand the use of the "ControlNet Input Directory" in the img2img batch tab, since I put it in the directory but nothing seemed to happen for me, but overall the other things work perfectly. 

thanks again.

ljleb commented:
As of right now, the text box in the img2img batch tab is used as a fallback directory. If you want all of your units to loop over the same images, you can simply use this field instead.

It will be removed as soon as we refactor the interface to include a separate accordion. See #1264 for where we want to take this next. I haven't took the time yet to look into doing the proposed change, there are too many things I am doing at the same time and I can't get a hold of my schedule... If anyone wants to take a bite at this, I'll be glad to review the code though.

p0mad commented:
@ljleb 
Thank you so much for the feature.

I was wondering if it is possible to input batch of pose(openpose+hands+face) and as a base output i want to initiate image or video of an animate character to generate video with the poses from input and the character?

I'm not sure if this is possible and how should i progress on that?

Best regards
